beacon-light-client/plonky2/crates/circuits/src/final_layer/mod.rs

```diff
@@ -0,0 +1,461 @@
+            balance_verification_proof_target: BasicProofTarget {
+                proof: balance_proof_targets.clone(),
+                verifier_circuit_target: balance_verifier_circuit_target,
+            },
+            commitment_mapper_proof_target: BasicProofTarget {
+                proof: commitment_mapper_proof_targets,
+                verifier_circuit_target: commitment_mapper_verifier_circuit_target,
+            },
+        }
+    }
+}
+
+fn setup_balance_targets<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+    const VALIDATORS_COUNT: usize,
+    const WITHDRAWAL_CREDENTIALS_COUNT: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    data: &CircuitData<F, C, D>,
+) -> (
+    ProofWithPublicInputsTarget<D>,
+    VerifierCircuitTarget,
+    CircuitOutputTarget<
+        WithdrawalCredentialsBalanceAggregatorFirstLevel<
+            VALIDATORS_COUNT,
+            WITHDRAWAL_CREDENTIALS_COUNT,
+        >,
+    >,
+)
+where
+    [(); VALIDATORS_COUNT / 4]:,
+    <C as GenericConfig<D>>::Hasher: AlgebraicHasher<F>,
+{
+    let (proof_targets, verifier_circuit_target) = setup_proof_targets(data, builder);
+
+    let output = WithdrawalCredentialsBalanceAggregatorFirstLevel::<
+        VALIDATORS_COUNT,
+        WITHDRAWAL_CREDENTIALS_COUNT,
+    >::read_public_inputs_target(&proof_targets.public_inputs);
+
+    (proof_targets, verifier_circuit_target, output)
+}
+
+fn setup_commitment_mapper_targets<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    data: &CircuitData<F, C, D>,
+) -> (
+    ProofWithPublicInputsTarget<D>,
+    VerifierCircuitTarget,
+    CircuitOutputTarget<ValidatorsCommitmentMapperFirstLevel>,
+)
+where
+    <C as GenericConfig<D>>::Hasher: AlgebraicHasher<F>,
+{
+    let (proof_targets, verifier_circuit_target) = setup_proof_targets(data, builder);
+
+    let output = ValidatorsCommitmentMapperFirstLevel::read_public_inputs_target(
+        &proof_targets.public_inputs,
+    );
+
+    (proof_targets, verifier_circuit_target, output)
+}
+
+// TODO: Rename this function
+fn verify_slot_is_in_range<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    slot: &BigUintTarget,
+    current_epoch: &BigUintTarget,
+) -> () {
+    let slots_per_epoch = builder.constant_biguint(&BigUint::from_u32(32).unwrap());
+    let slot_epoch = builder.div_biguint(slot, &slots_per_epoch);
+    builder.connect_biguint(&slot_epoch, current_epoch);
+}
+
+fn setup_proof_targets<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+    circuit_data: &CircuitData<F, C, D>,
+    builder: &mut CircuitBuilder<F, D>,
+) -> (ProofWithPublicInputsTarget<D>, VerifierCircuitTarget)
+where
+    <C as GenericConfig<D>>::Hasher: AlgebraicHasher<F>,
+{
+    let verifier_circuit_target = VerifierCircuitTarget {
```

Comment:
I think we should use constant_merkle_cap and constant_hash as we do on the other places
Cause I am not sure if this isn't a vulnerability actually having in mind that we don't pass the verifiercircuit target as public_input and we don't validate it afterwards I think it would be possible to recursively verify any circuit in this final layer

Reply:
It was like that before the refactoring just lets fix it here